### ISSUE TITLE
Failed to find pointer-size integer type on Alpine Linux / musl

### ISSUE THREAD

**Issue by JoeUX** [madler/zlib#246]
Hello, I'm running into an obstacle running `./configure` on Alpine Linux v. 3.5.2 (latest), x86_64 version. The error says "Failed to find a pointer-size integer type", whereupon it aborts. Screenshot below:

![2017-04-01_20-50-21](https://cloud.githubusercontent.com/assets/8071692/24584416/8696b506-1722-11e7-8921-0af9797e19ce.png)

Alpine Linux uses the [musl c library](https://www.musl-libc.org/) instead of glibc. Is this part of the issue? How can I fix this?

Other info: I doubt this would matter, but I'm running Alpine as a guest in Virtual Box 5.1.18 on a Windows 10 host.

I'm following the [setup instructions](https://www.nginx.com/resources/admin-guide/installing-nginx-open-source/) on the nginx website, but at this stage nothing nginx-related is happening yet.

**Comment by madler** [comment]
Please do a `make distclean` followed by a `./configure`, and attach the resulting `configure.log` to this issue. Thanks.

**Comment by JoeArizona** [comment]
Sorry for the delay. Here it is:

```
./configure
Sun Apr 16 17:25:35 MST 2017
Checking for gcc...
=== ztest1857.c ===
extern int getchar();
int hello() {return getchar();}
===
gcc -c ztest1857.c
... using gcc

Checking for obsessive-compulsive compiler options...
=== ztest1857.c ===
int foo() { return 0; }
===
gcc -c -O3 ztest1857.c

Checking for shared library support...
=== ztest1857.c ===
extern int getchar();
int hello() {return getchar();}
===
gcc -w -c -O3 -fPIC ztest1857.c
gcc -shared -Wl,-soname,libz.so.1,--version-script,zlib.map -O3 -fPIC -o ztest18
/usr/lib/gcc/x86_64-alpine-linux-musl/6.2.1/../../../../x86_64-alpine-linux-musl
/usr/lib/gcc/x86_64-alpine-linux-musl/6.2.1/../../../../x86_64-alpine-linux-musl
collect2: error: ld returned 1 exit status
(exit code 1)
No shared library support.
Building static library libz.a version 1.2.11 with gcc.

=== ztest1857.c ===
#include <stdio.h>
#include <stdlib.h>
size_t dummy = 0;
===
gcc -c -O3 ztest1857.c
ztest1857.c:1:19: fatal error: stdio.h: No such file or directory
 #include <stdio.h>
                   ^
compilation terminated.
(exit code 1)
Checking for size_t... No.

=== ztest1857.c ===
long long dummy = 0;
===
gcc -c -O3 ztest1857.c
Checking for long long... Yes.
=== ztest1857.c ===
#include <stdio.h>
int main(void) {
    if (sizeof(void *) <= sizeof(int)) puts("int");
    else if (sizeof(void *) <= sizeof(long)) puts("long");
    else puts("z_longlong");
    return 0;
}
===
gcc -O3 -o ztest1857 ztest1857.c
ztest1857.c:1:19: fatal error: stdio.h: No such file or directory
 #include <stdio.h>
                   ^
compilation terminated.
(exit code 1)
Failed to find a pointer-size integer type.
** ./configure aborting.
--------------------
```

**Comment by adyshimony** [comment]
@JoeUX 
Did you find what cause the error? I am stuck with it. 
Thanks.

**Comment by haoxinyuan** [comment]
> @JoeUX
> Did you find what cause the error? I am stuck with it.
> Thanks.

Have u solved。I have the same issue。
Thanks.

**Comment by haoxinyuan** [comment]
execute yum install -y gcc gcc-c++ glibc zlib-devel openssl-devel pam-devel
I have solved

**Comment by JoeUX** [comment]
@adyshimony @haoxinyuan I never figured it out. I'll try @haoxinyuan's new fix.

**Comment by madler** [comment]
```
ztest1857.c:1:19: fatal error: stdio.h: No such file or directory
 #include <stdio.h>
```

If you can't include stdio.h, then your compiler installation just plain doesn't work.

**Comment by iav** [comment]
On alpine linux `apk add libc-dev` install standart headers.

**Comment by Duyj0109** [comment]
@JoeUX 
我遇到跟您一样的问题，但是我没有办法通过yum进行操作，您有什么好的解决方案么？

**Comment by JoeUX** [comment]
Okay, I've tested the different ideas. The solution posted by @iav works. That's `apk add libc-dev`. Note that this also installs musl-dev. I don't know whether it's libc-dev or musl-dev that fixes the problem, but I assume it's libc-dev.

@haoxinyuan Your solution didn't quite work. Note that this issue is about configuring zlib with nginx on **Alpine Linux**, not Ubuntu. Your syntax is for Ubuntu. The error is specific to Alpine Linux and its default C library header configuration. Alpine uses the musl C library by default, not glibc that most Linux distros use. glibc isn't available from the Alpine package repo, so it's not possible to run your command. Also, the packages you referenced have different names on Ubuntu and Alpine. zlib-devel is called zlib-dev on Alpine, for example. I tried installing zlib-dev, but it didn't fix the problem – same error. Installing libc-dev fixed it.
